<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>FAST: simulation • FAST</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="FAST: simulation">
<meta property="og:description" content="FAST">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">FAST</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/FASTdlpfc2.html">FAST: two DLPFC sections</a>
    </li>
    <li>
      <a href="../articles/FASTsimu.html">FAST: simulation</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/feiyoung/FAST/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>FAST: simulation</h1>
                        <h4 data-toc-skip class="author">Wei Liu</h4>
            
            <h4 data-toc-skip class="date">2023-07-28</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/feiyoung/FAST/blob/HEAD/vignettes/FASTsimu.Rmd" class="external-link"><code>vignettes/FASTsimu.Rmd</code></a></small>
      <div class="hidden name"><code>FASTsimu.Rmd</code></div>

    </div>

    
    
<p>This vignette introduces the FAST workflow for the analysis of multiple simulated spatial transcriptomics dataset. FAST workflow is based on the PRECASTObj object created in the <code>PRECAST</code> R package and the workflow of FAST is similar to that of PRECAST; see (<a href="https://feiyoung.github.io/PRECAST/articles/PRECAST.BreastCancer.html" class="external-link uri">https://feiyoung.github.io/PRECAST/articles/PRECAST.BreastCancer.html</a>) for the workflow of PRECAST. The workflow of FAST consists of three steps</p>
<ul>
<li>Independent preprocessing and model setting</li>
<li>Spatial dimension reduction using FAST</li>
<li>Downstream analysis (i.e. , embedding alignment using Harmony and clustering using Louvain, visualization of clusters and embeddings, remove unwanted variation, combined differential expression analysis)</li>
<li>Except for the above downstream analyses, cell-cell interaction analysis and trajectory inference can also be performed on the basis of the embeddings obtained by FAST.</li>
</ul>
<p>We demonstrate the use of FAST to three simulated ST data that are <a href="https://github.com/feiyoung/FAST/tree/main/vignettes_data" class="external-link">here</a>, which can be downloaded to the current working path by the following command:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">githubURL</span> <span class="op">&lt;-</span> <span class="st">"https://github.com/feiyoung/FAST/blob/main/vignettes_data/simu3.rds?raw=true"</span>
<span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="va">githubURL</span>,<span class="st">"simu3.rds"</span>,mode<span class="op">=</span><span class="st">'wb'</span><span class="op">)</span></code></pre></div>
<p>Then load to R</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="st">"simu3.rds"</span><span class="op">)</span></code></pre></div>
<p>The package can be loaded with the command:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/feiyoung/FAST" class="external-link">FAST</a></span><span class="op">)</span>
<span class="co">#&gt; Loading required package: gtools</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; </span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/feiyoung/PRECAST" class="external-link">PRECAST</a></span><span class="op">)</span>
<span class="co">#&gt; Loading required package: parallel</span>
<span class="co">#&gt; PRECAST :  An efficient data integration method is provided for multiple spatial transcriptomics data with non-cluster-relevant effects such as the complex batch effects. It unifies spatial factor analysis simultaneously with spatial clustering and embedding alignment, requiring only partially shared cell/domain clusters across datasets. More details can be referred to Wei Liu, et al. (2023) &lt;doi:10.1038/s41467-023-35947-w&gt;.   Check out our Package website (https://feiyoung.github.io/PRECAST/index.html) for a more complete description of the methods and analyses</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span>
<span class="co">#&gt; Warning: package 'Seurat' was built under R version 4.1.3</span>
<span class="co">#&gt; Attaching SeuratObject</span>
<span class="co">#&gt; Attaching sp</span></code></pre></div>
<div class="section level2">
<h2 id="view-the-simulated-data">View the simulated data<a class="anchor" aria-label="anchor" href="#view-the-simulated-data"></a>
</h2>
<p>First, we view the the three simulated spatial transcriptomics data with ST platform. There are 200 genes for each data batch and ~2000 spots in total</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">simu3</span> <span class="co">## a list including three Seurat object with default assay: RNA</span>
<span class="co">#&gt; [[1]]</span>
<span class="co">#&gt; An object of class Seurat </span>
<span class="co">#&gt; 200 features across 625 samples within 1 assay </span>
<span class="co">#&gt; Active assay: RNA (200 features, 0 variable features)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; [[2]]</span>
<span class="co">#&gt; An object of class Seurat </span>
<span class="co">#&gt; 200 features across 400 samples within 1 assay </span>
<span class="co">#&gt; Active assay: RNA (200 features, 0 variable features)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; [[3]]</span>
<span class="co">#&gt; An object of class Seurat </span>
<span class="co">#&gt; 200 features across 900 samples within 1 assay </span>
<span class="co">#&gt; Active assay: RNA (200 features, 0 variable features)</span></code></pre></div>
<p>Check the content in <code>simu3</code>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">simu3</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/row.names.html" class="external-link">row.names</a></span><span class="op">(</span><span class="va">simu3</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="create-a-precastobject-object">Create a PRECASTObject object<a class="anchor" aria-label="anchor" href="#create-a-precastobject-object"></a>
</h2>
<p>We show how to create a PRECASTObject object step by step. First, we create a Seurat list object using the count matrix and meta data of each data batch. Although <code>simu3</code> is a prepared Seurat list object, we re-create a same objcet seuList to show the details.</p>
<ul>
<li>Note: the spatial coordinates must be contained in the meta data and named as <code>row</code> and <code>col</code>, which benefits the identification of spaital coordinates by FAST</li>
</ul>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Get the gene-by-spot read count matrices</span>
<span class="va">countList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">simu3</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="va">x</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span><span class="op">@</span><span class="va">counts</span><span class="op">)</span>

<span class="co">## Check the spatial coordinates: Yes, they are named as "row" and "col"!</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">simu3</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">@</span><span class="va">meta.data</span><span class="op">)</span>
<span class="co">#&gt;          orig.ident nCount_RNA nFeature_RNA row col  Group</span>
<span class="co">#&gt; spot1 SeuratProject     130934          114   1   1 Layer4</span>
<span class="co">#&gt; spot2 SeuratProject       3103          116   2   1 Layer6</span>
<span class="co">#&gt; spot3 SeuratProject       8112          110   3   1 Layer6</span>
<span class="co">#&gt; spot4 SeuratProject       2637          105   4   1 Layer6</span>
<span class="co">#&gt; spot5 SeuratProject       5664          109   5   1 Layer5</span>
<span class="co">#&gt; spot6 SeuratProject       4636          107   6   1 Layer6</span>

<span class="co">## Get the meta data of each spot for each data batch</span>
<span class="va">metadataList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">simu3</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="va">x</span><span class="op">@</span><span class="va">meta.data</span><span class="op">)</span>


<span class="co">## ensure the row.names of metadata in metaList are the same as that of colnames count matrix in countList</span>
<span class="va">M</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">countList</span><span class="op">)</span>
<span class="kw">for</span><span class="op">(</span><span class="va">r</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="va">M</span><span class="op">)</span><span class="op">{</span>
  <span class="fu"><a href="https://rdrr.io/r/base/row.names.html" class="external-link">row.names</a></span><span class="op">(</span><span class="va">metadataList</span><span class="op">[[</span><span class="va">r</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">countList</span><span class="op">[[</span><span class="va">r</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>
<span class="op">}</span>


<span class="co">## Create the Seurat list  object</span>

<span class="va">seuList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span>
<span class="kw">for</span><span class="op">(</span><span class="va">r</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="va">M</span><span class="op">)</span><span class="op">{</span>
  <span class="va">seuList</span><span class="op">[[</span><span class="va">r</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">countList</span><span class="op">[[</span><span class="va">r</span><span class="op">]</span><span class="op">]</span>, meta.data<span class="op">=</span><span class="va">metadataList</span><span class="op">[[</span><span class="va">r</span><span class="op">]</span><span class="op">]</span>, project <span class="op">=</span> <span class="st">"FASTsimu"</span><span class="op">)</span>
<span class="op">}</span></code></pre></div>
<div class="section level3">
<h3 id="prepare-the-precastobject-with-preprocessing-step-">Prepare the PRECASTObject with preprocessing step.<a class="anchor" aria-label="anchor" href="#prepare-the-precastobject-with-preprocessing-step-"></a>
</h3>
<p>Next, we use <code><a href="https://rdrr.io/pkg/PRECAST/man/CreateiDRSCObject.html" class="external-link">CreatePRECASTObject()</a></code> to create a PRECASTObject object based on the Seurat list object <code>seuList</code>. Users are able to see <a href="https://feiyoung.github.io/PRECAST/articles/PRECAST.BreastCancer.html" class="external-link uri">https://feiyoung.github.io/PRECAST/articles/PRECAST.BreastCancer.html</a> for what is done in this function. Since this is a simulated dataset, we used all the 200 genes by using a custom gene list <code>customGenelist=custom_genelist)</code>. We observe that there are only 197 genes passing the filtering step.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="co">## Create PRECASTObject</span>
<span class="va">custom_genelist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/row.names.html" class="external-link">row.names</a></span><span class="op">(</span><span class="va">seuList</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">2023</span><span class="op">)</span>
<span class="va">PRECASTObj</span> <span class="op">&lt;-</span>  <span class="fu"><a href="https://rdrr.io/pkg/PRECAST/man/CreateiDRSCObject.html" class="external-link">CreatePRECASTObject</a></span><span class="op">(</span><span class="va">seuList</span>, customGenelist<span class="op">=</span><span class="va">custom_genelist</span><span class="op">)</span>
<span class="co">#&gt; Filter spots and features from Raw count data...</span>
<span class="co">#&gt;  </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; CreatePRECASTObject: remove genes:gene90  gene144  gene170  with low count reads in seuList.</span>
<span class="co">#&gt; Filter spots and features from SVGs(HVGs) count data...</span>

<span class="co">## User can retain the raw seuList by the following commond.</span>
<span class="co">##  PRECASTObj &lt;-  CreatePRECASTObject(seuList, customGenelist=row.names(seuList[[1]]), rawData.preserve = TRUE)</span>
<span class="co">## check the number of genes/features after filtering step</span>
<span class="va">PRECASTObj</span><span class="op">@</span><span class="va">seulist</span>
<span class="co">#&gt; [[1]]</span>
<span class="co">#&gt; An object of class Seurat </span>
<span class="co">#&gt; 197 features across 625 samples within 1 assay </span>
<span class="co">#&gt; Active assay: RNA (197 features, 0 variable features)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; [[2]]</span>
<span class="co">#&gt; An object of class Seurat </span>
<span class="co">#&gt; 197 features across 400 samples within 1 assay </span>
<span class="co">#&gt; Active assay: RNA (197 features, 0 variable features)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; [[3]]</span>
<span class="co">#&gt; An object of class Seurat </span>
<span class="co">#&gt; 197 features across 900 samples within 1 assay </span>
<span class="co">#&gt; Active assay: RNA (197 features, 0 variable features)</span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="fit-precast-using-simulated-data">Fit PRECAST using simulated data<a class="anchor" aria-label="anchor" href="#fit-precast-using-simulated-data"></a>
</h2>
<div class="section level3">
<h3 id="add-the-model-setting">Add the model setting<a class="anchor" aria-label="anchor" href="#add-the-model-setting"></a>
</h3>
<p>Add adjacency matrix list and parameter setting of FAST. More model setting parameters can be found in <code><a href="../reference/model_set_FAST.html">model_set_FAST()</a></code>.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="co">## seuList is null since the default value `rawData.preserve` is FALSE.</span>
<span class="va">PRECASTObj</span><span class="op">@</span><span class="va">seuList</span>
<span class="co">#&gt; NULL</span>

<span class="co">## Add adjacency matrix list for a PRECASTObj object to prepare for FAST model fitting.</span>
<span class="va">PRECASTObj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/PRECAST/man/AddAdjList.html" class="external-link">AddAdjList</a></span><span class="op">(</span><span class="va">PRECASTObj</span>, platform <span class="op">=</span> <span class="st">"ST"</span><span class="op">)</span>
<span class="co">#&gt; Neighbors were identified for 625 out of 625 spots.</span>
<span class="co">#&gt; Neighbors were identified for 400 out of 400 spots.</span>
<span class="co">#&gt; Neighbors were identified for 900 out of 900 spots.</span>

<span class="co">## Add a model setting in advance for a PRECASTObj object: verbose =TRUE helps outputing the information in the algorithm; </span>
<span class="va">PRECASTObj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AddParSettingFAST.html">AddParSettingFAST</a></span><span class="op">(</span><span class="va">PRECASTObj</span>, verbose<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>
<span class="co">## Check the parameters</span>
<span class="va">PRECASTObj</span><span class="op">@</span><span class="va">parameterList</span>
<span class="co">#&gt; $maxIter</span>
<span class="co">#&gt; [1] 30</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $seed</span>
<span class="co">#&gt; [1] 1</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $epsLogLik</span>
<span class="co">#&gt; [1] 1e-05</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $verbose</span>
<span class="co">#&gt; [1] TRUE</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $error_heter</span>
<span class="co">#&gt; [1] TRUE</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $Psi_diag</span>
<span class="co">#&gt; [1] FALSE</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="fit-fast">Fit FAST<a class="anchor" aria-label="anchor" href="#fit-fast"></a>
</h3>
<p>For function <code>FAST</code>, users can specify the number of factors <code>q</code> and the fitted model <code>fit.model</code>. The <code>q</code> sets the number of spatial factors to be extracted, and a lareger one means more information to be extracted but higher computaional cost. The <code>fit.model</code> specifies the version of FAST to be fitted. The Gaussian version (<code>gaussian</code>) models the log-count matrices while the Poisson verion (<code>poisson</code>) models the count matrices; default as <code>poisson</code>.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">### set q= 20 here</span>
<span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">2023</span><span class="op">)</span>
<span class="va">PRECASTObj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/FAST.html">FAST</a></span><span class="op">(</span><span class="va">PRECASTObj</span>, q<span class="op">=</span><span class="fl">20</span><span class="op">)</span>
<span class="co">#&gt; Warning in if (!(fit.model %in% c("poisson", "gaussian"))) stop("FAST: Check the</span>
<span class="co">#&gt; argument: fit.model! fit.model must either be 'poisson' or 'gaussian'."): the</span>
<span class="co">#&gt; condition has length &gt; 1 and only the first element will be used</span>
<span class="co">#&gt; ******Run the Poisson version of FAST...</span>
<span class="co">#&gt; Finish variable initialization </span>
<span class="co">#&gt; Satrt ICM step! </span>
<span class="co">#&gt; iter = 2, ELBO= 189481969.596825, dELBO=1.088234 </span>
<span class="co">#&gt; Satrt ICM step! </span>
<span class="co">#&gt; iter = 3, ELBO= 208013944.617564, dELBO=0.097803 </span>
<span class="co">#&gt; Satrt ICM step! </span>
<span class="co">#&gt; iter = 4, ELBO= 208041974.387419, dELBO=0.000135 </span>
<span class="co">#&gt; Satrt ICM step! </span>
<span class="co">#&gt; iter = 5, ELBO= 208057832.422356, dELBO=0.000076 </span>
<span class="co">#&gt; Satrt ICM step! </span>
<span class="co">#&gt; iter = 6, ELBO= 208067869.378179, dELBO=0.000048 </span>
<span class="co">#&gt; Satrt ICM step! </span>
<span class="co">#&gt; iter = 7, ELBO= 208074731.784015, dELBO=0.000033 </span>
<span class="co">#&gt; Satrt ICM step! </span>
<span class="co">#&gt; iter = 8, ELBO= 208079699.135512, dELBO=0.000024 </span>
<span class="co">#&gt; Satrt ICM step! </span>
<span class="co">#&gt; iter = 9, ELBO= 208083457.759797, dELBO=0.000018 </span>
<span class="co">#&gt; Satrt ICM step! </span>
<span class="co">#&gt; iter = 10, ELBO= 208086403.143739, dELBO=0.000014 </span>
<span class="co">#&gt; Satrt ICM step! </span>
<span class="co">#&gt; iter = 11, ELBO= 208088775.191483, dELBO=0.000011 </span>
<span class="co">#&gt; Satrt ICM step! </span>
<span class="co">#&gt; iter = 12, ELBO= 208090727.539180, dELBO=0.000009</span>
<span class="co">#&gt; 2023-07-28 09:35:48 : ***** Finish FAST, 0.051 mins elapsed.</span>
<span class="co">### Check the results</span>
<span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">PRECASTObj</span><span class="op">@</span><span class="va">resList</span><span class="op">)</span>
<span class="co">#&gt; List of 1</span>
<span class="co">#&gt;  $ FAST:List of 7</span>
<span class="co">#&gt;   ..$ hV      :List of 3</span>
<span class="co">#&gt;   .. ..$ : num [1:625, 1:15] -2.27 -4.05 -6.52 -1.77 -1.74 ...</span>
<span class="co">#&gt;   .. .. ..- attr(*, "scaled:center")= num [1:15] -0.867 -0.848 -2.528 0.32 1.01 ...</span>
<span class="co">#&gt;   .. ..$ : num [1:400, 1:15] -9.32 -3.48 -1.82 -5.13 -7.25 ...</span>
<span class="co">#&gt;   .. .. ..- attr(*, "scaled:center")= num [1:15] -0.0347 0.7836 -0.8125 -0.9521 3.9392 ...</span>
<span class="co">#&gt;   .. ..$ : num [1:900, 1:15] 1.019 -2.055 -0.624 -5.129 -2.433 ...</span>
<span class="co">#&gt;   .. .. ..- attr(*, "scaled:center")= num [1:15] -3.627 -0.795 1.606 1.231 -3.135 ...</span>
<span class="co">#&gt;   ..$ nu      : num [1:3, 1:197] 1.037 4.191 -1.12 -0.305 -0.44 ...</span>
<span class="co">#&gt;   ..$ Psi     : num [1:15, 1:15, 1:3] 29.685 3.493 -0.744 -2.433 -0.368 ...</span>
<span class="co">#&gt;   ..$ W       : num [1:197, 1:15] 0.16046 -0.07676 -0.08422 -0.00717 -0.03362 ...</span>
<span class="co">#&gt;   ..$ Lam     : num [1:3, 1:197] 1.84 2.4 1.14 1.61 1.12 ...</span>
<span class="co">#&gt;   ..$ ELBO    : num 2.08e+08</span>
<span class="co">#&gt;   ..$ ELBO_seq: num [1:11] -2.15e+09 1.89e+08 2.08e+08 2.08e+08 2.08e+08 ...</span></code></pre></div>
<details><p><summary><strong>Run gaussian version</strong></summary> Users can also use the gaussian version by the following command:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">2023</span><span class="op">)</span>
<span class="va">PRECASTObj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/FAST.html">FAST</a></span><span class="op">(</span><span class="va">PRECASTObj</span>, q<span class="op">=</span><span class="fl">20</span>, fit.model<span class="op">=</span><span class="st">'gaussian'</span><span class="op">)</span>
<span class="co">### Check the results</span>
<span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">PRECASTObj</span><span class="op">@</span><span class="va">resList</span><span class="op">)</span></code></pre></div>
</details>
</div>
<div class="section level3">
<h3 id="evaluate-the-adjusted-mcfaddens-pseudo-r-square">Evaluate the adjusted McFadden’s pseudo R-square<a class="anchor" aria-label="anchor" href="#evaluate-the-adjusted-mcfaddens-pseudo-r-square"></a>
</h3>
<p>Next, we investigate the performance of dimension reduction by calculating the adjusted McFadden’s pseudo R-square for each data batch. The simulated true labels is in the <code>meta.data</code> of each component of <code>PRECASTObj@seulist</code>.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Obtain the true labels</span>
<span class="va">yList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">PRECASTObj</span><span class="op">@</span><span class="va">seulist</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="va">x</span><span class="op">$</span><span class="va">Group</span><span class="op">)</span>
<span class="co">### Evaluate the MacR2</span>
<span class="va">MacVec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">PRECASTObj</span><span class="op">@</span><span class="va">seulist</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">r</span><span class="op">)</span> <span class="fu"><a href="../reference/get_r2_mcfadden.html">get_r2_mcfadden</a></span><span class="op">(</span><span class="va">PRECASTObj</span><span class="op">@</span><span class="va">resList</span><span class="op">$</span><span class="va">FAST</span><span class="op">$</span><span class="va">hV</span><span class="op">[[</span><span class="va">r</span><span class="op">]</span><span class="op">]</span>, <span class="va">yList</span><span class="op">[[</span><span class="va">r</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; # weights:  119 (96 variable)</span>
<span class="co">#&gt; initial  value 1216.193843 </span>
<span class="co">#&gt; iter  10 value 114.780227</span>
<span class="co">#&gt; iter  20 value 92.092862</span>
<span class="co">#&gt; iter  30 value 85.341446</span>
<span class="co">#&gt; iter  40 value 82.688874</span>
<span class="co">#&gt; iter  50 value 81.299335</span>
<span class="co">#&gt; iter  60 value 79.578069</span>
<span class="co">#&gt; iter  70 value 77.914256</span>
<span class="co">#&gt; iter  80 value 77.803555</span>
<span class="co">#&gt; iter  90 value 77.775020</span>
<span class="co">#&gt; iter 100 value 77.773307</span>
<span class="co">#&gt; final  value 77.773307 </span>
<span class="co">#&gt; stopped after 100 iterations</span>
<span class="co">#&gt; # weights:  119 (96 variable)</span>
<span class="co">#&gt; initial  value 778.364060 </span>
<span class="co">#&gt; iter  10 value 116.386127</span>
<span class="co">#&gt; iter  20 value 28.870844</span>
<span class="co">#&gt; iter  30 value 7.501395</span>
<span class="co">#&gt; iter  40 value 3.686866</span>
<span class="co">#&gt; iter  50 value 1.472696</span>
<span class="co">#&gt; iter  60 value 0.878809</span>
<span class="co">#&gt; iter  70 value 0.613061</span>
<span class="co">#&gt; iter  80 value 0.417069</span>
<span class="co">#&gt; iter  90 value 0.287838</span>
<span class="co">#&gt; iter 100 value 0.159742</span>
<span class="co">#&gt; final  value 0.159742 </span>
<span class="co">#&gt; stopped after 100 iterations</span>
<span class="co">#&gt; # weights:  119 (96 variable)</span>
<span class="co">#&gt; initial  value 1751.319134 </span>
<span class="co">#&gt; iter  10 value 147.392347</span>
<span class="co">#&gt; iter  20 value 126.553241</span>
<span class="co">#&gt; iter  30 value 119.448945</span>
<span class="co">#&gt; iter  40 value 118.722384</span>
<span class="co">#&gt; iter  50 value 117.568144</span>
<span class="co">#&gt; iter  60 value 115.680333</span>
<span class="co">#&gt; iter  70 value 113.602190</span>
<span class="co">#&gt; iter  80 value 113.298580</span>
<span class="co">#&gt; iter  90 value 113.270105</span>
<span class="co">#&gt; iter 100 value 113.254185</span>
<span class="co">#&gt; final  value 113.254185 </span>
<span class="co">#&gt; stopped after 100 iterations</span>
<span class="co">### output them</span>
<span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">MacVec</span><span class="op">)</span>
<span class="co">#&gt; adjusted McFadden's R2 adjusted McFadden's R2 adjusted McFadden's R2 </span>
<span class="co">#&gt;              0.9348610              0.9984897              0.9343897</span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="embedding-alignment-and-clustering-using-harmony-and-louvain">Embedding alignment and clustering using Harmony and Louvain<a class="anchor" aria-label="anchor" href="#embedding-alignment-and-clustering-using-harmony-and-louvain"></a>
</h2>
<p>Based on the embeddings from FAST, we use <code>Harmony</code> to align the embeddings then followed by Louvain clustering to obtain the cluster labels. In this downstream analysis, other methods for embedding alignment and clustering can be also used. In the vignette of two sections of DLPFC Visium data, we will show another method (<code>iSC-MEB</code>) to jointly perform embedding alignment and spatial clustering.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">PRECASTObj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RunHarmonyLouvain.html">RunHarmonyLouvain</a></span><span class="op">(</span><span class="va">PRECASTObj</span>, resolution <span class="op">=</span> <span class="fl">0.4</span><span class="op">)</span>
<span class="co">#&gt; ******Use Harmony to remove batch in the embeddings from FAST...</span>
<span class="co">#&gt; Harmony 1/10</span>
<span class="co">#&gt; Harmony 2/10</span>
<span class="co">#&gt; Harmony 3/10</span>
<span class="co">#&gt; Harmony 4/10</span>
<span class="co">#&gt; Harmony 5/10</span>
<span class="co">#&gt; Harmony converged after 5 iterations</span>
<span class="co">#&gt; 2023-07-28 09:35:51 : ***** Finish Harmony correction, 0.043 mins elapsed.</span>
<span class="co">#&gt; ******Use Louvain to cluster and determine the number of clusters...</span>
<span class="co">#&gt; Warning: No assay specified, setting assay as RNA by default.</span>
<span class="co">#&gt; Warning: Keys should be one or more alphanumeric characters followed by an</span>
<span class="co">#&gt; underscore, setting key from PC to PC_</span>
<span class="co">#&gt; Warning: All keys should be one or more alphanumeric characters followed by an</span>
<span class="co">#&gt; underscore '_', setting key to PC_</span>
<span class="co">#&gt; Computing nearest neighbor graph</span>
<span class="co">#&gt; Computing SNN</span>
<span class="co">#&gt; Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Number of nodes: 1925</span>
<span class="co">#&gt; Number of edges: 83023</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Running Louvain algorithm...</span>
<span class="co">#&gt; Maximum modularity in 10 random starts: 0.8788</span>
<span class="co">#&gt; Number of communities: 7</span>
<span class="co">#&gt; Elapsed time: 0 seconds</span>
<span class="co">#&gt; 2023-07-28 09:35:52 : ***** Finsh Louvain clustering and find the optimal number of clusters, 0.018 mins elapsed.</span>
<span class="va">ARI_vec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="va">M</span>, <span class="kw">function</span><span class="op">(</span><span class="va">r</span><span class="op">)</span> <span class="fu">mclust</span><span class="fu">::</span><span class="fu"><a href="https://mclust-org.github.io/mclust/reference/adjustedRandIndex.html" class="external-link">adjustedRandIndex</a></span><span class="op">(</span><span class="va">PRECASTObj</span><span class="op">@</span><span class="va">resList</span><span class="op">$</span><span class="va">Louvain</span><span class="op">$</span><span class="va">cluster</span><span class="op">[[</span><span class="va">r</span><span class="op">]</span><span class="op">]</span>, <span class="va">yList</span><span class="op">[[</span><span class="va">r</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">ARI_vec</span><span class="op">)</span>
<span class="co">#&gt; [1] 0.8315672 0.8625198 0.8126047</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="remove-unwanted-variations-in-the-log-normalized-expression-matrices">Remove unwanted variations in the log-normalized expression matrices<a class="anchor" aria-label="anchor" href="#remove-unwanted-variations-in-the-log-normalized-expression-matrices"></a>
</h2>
<p>In the following, we remove the unwanted variations in the log-normalized expression matrices to obtain a combined log-normalized expression matrix in a Seurat object. In the context of the simulated data used in this study, housekeeping genes are not present, thus, we turn to another method to remove the unwanted variations. Specifically, we leverage the batch effect embeddings estimated through Harmony to capture and mitigate unwanted variations. Additionally, we utilize the cluster labels obtained via Louvain clustering to retain the desired biological effects.</p>
<p>The estimated embeddings of batch effects (batchEmbed) are in the slot <code>PRECASTObj@resList$Harmony</code> and cluster labels (cluster) are in the slot <code>PRECASTObj@resList$Louvain</code>.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">PRECASTObj</span><span class="op">@</span><span class="va">resList</span><span class="op">$</span><span class="va">Harmony</span><span class="op">)</span>
<span class="co">#&gt; List of 2</span>
<span class="co">#&gt;  $ harmonyembed:List of 3</span>
<span class="co">#&gt;   ..$ : num [1:625, 1:15] -2.19 -4.13 -6.67 -1.75 -2.28 ...</span>
<span class="co">#&gt;   ..$ : num [1:400, 1:15] -8.96 -3.13 -1.87 -4.88 -6.04 ...</span>
<span class="co">#&gt;   ..$ : num [1:900, 1:15] 1.143 -2.349 -0.894 -5.486 -2.931 ...</span>
<span class="co">#&gt;  $ batchEmbed  :List of 3</span>
<span class="co">#&gt;   ..$ : num [1:625, 1:15] -0.24035 0.00474 -0.02522 -0.05699 0.33849 ...</span>
<span class="co">#&gt;   .. ..- attr(*, "dimnames")=List of 2</span>
<span class="co">#&gt;   .. .. ..$ : chr [1:625] "1" "2" "3" "4" ...</span>
<span class="co">#&gt;   .. .. ..$ : NULL</span>
<span class="co">#&gt;   ..$ : num [1:400, 1:15] -0.4123 -0.4013 -0.0554 -0.3333 -1.3548 ...</span>
<span class="co">#&gt;   .. ..- attr(*, "dimnames")=List of 2</span>
<span class="co">#&gt;   .. .. ..$ : chr [1:400] "626" "627" "628" "629" ...</span>
<span class="co">#&gt;   .. .. ..$ : NULL</span>
<span class="co">#&gt;   ..$ : num [1:900, 1:15] -0.0506 0.1896 0.1548 0.0342 0.3375 ...</span>
<span class="co">#&gt;   .. ..- attr(*, "dimnames")=List of 2</span>
<span class="co">#&gt;   .. .. ..$ : chr [1:900] "1026" "1027" "1028" "1029" ...</span>
<span class="co">#&gt;   .. .. ..$ : NULL</span>
<span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">PRECASTObj</span><span class="op">@</span><span class="va">resList</span><span class="op">$</span><span class="va">Louvain</span><span class="op">)</span>
<span class="co">#&gt; List of 1</span>
<span class="co">#&gt;  $ cluster:List of 3</span>
<span class="co">#&gt;   ..$ : Factor w/ 7 levels "0","1","2","3",..: 6 2 2 2 3 2 4 1 2 5 ...</span>
<span class="co">#&gt;   .. ..- attr(*, "names")= chr [1:625] "spot1" "spot2" "spot3" "spot4" ...</span>
<span class="co">#&gt;   ..$ : Factor w/ 7 levels "0","1","2","3",..: 2 5 4 2 3 3 3 3 1 1 ...</span>
<span class="co">#&gt;   .. ..- attr(*, "names")= chr [1:400] "spot626" "spot627" "spot628" "spot629" ...</span>
<span class="co">#&gt;   ..$ : Factor w/ 7 levels "0","1","2","3",..: 6 6 6 6 4 5 5 5 6 5 ...</span>
<span class="co">#&gt;   .. ..- attr(*, "names")= chr [1:900] "spot1026" "spot1027" "spot1028" "spot1029" ...</span></code></pre></div>
<p>Then, we integrate the three sections by removing the unwanted variations and setting <code>seulist_HK=NULL</code> and <code>Method = "HarmonyLouvain"</code> in the function <code><a href="../reference/IntegrateSRTData.html">IntegrateSRTData()</a></code>. After obtaining <code>seuInt</code>, we will see there are three embeddings: <code>profast</code>, <code>harmony</code> and <code>position</code>, in the slot <code>seuInt@reductions</code>. <code>profast</code> are the embeddings obtained by FAST model fitting and are uncorrected embeddings that may includes the unwanted effects (i.e., batch effects); <code>harmony</code>are the embeddings obtained by Harmony fitting and are aligned embeddings; and <code>position</code> are the spatial coordinates.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">seulist_HK</span> <span class="op">&lt;-</span> <span class="cn">NULL</span>
<span class="va">seuInt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/IntegrateSRTData.html">IntegrateSRTData</a></span><span class="op">(</span><span class="va">PRECASTObj</span>, seulist_HK<span class="op">=</span><span class="va">seulist_HK</span>, Method <span class="op">=</span> <span class="st">"HarmonyLouvain"</span>, seuList_raw<span class="op">=</span><span class="cn">NULL</span>, covariates_use<span class="op">=</span><span class="cn">NULL</span>, verbose<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>
<span class="co">#&gt; ******Perform PCA on housekeeping gene expression matrix...</span>
<span class="co">#&gt; Only use  the results in Harmony and Louvain to remove unwanted variations...</span>
<span class="co">#&gt; 2023-07-28 09:35:52 : ***** Finish PCA, 0.001 mins elapsed.</span>
<span class="co">#&gt; ******Remove the unwanted variations in gene expressions using spatial linear regression...</span>
<span class="co">#&gt; Finish variable initialization </span>
<span class="co">#&gt; iter = 2, ELBO= -506059.111341, dELBO=0.999764 </span>
<span class="co">#&gt; iter = 3, ELBO= -441635.737161, dELBO=0.127304 </span>
<span class="co">#&gt; iter = 4, ELBO= -415220.082023, dELBO=0.059813 </span>
<span class="co">#&gt; iter = 5, ELBO= -396081.727921, dELBO=0.046092 </span>
<span class="co">#&gt; iter = 6, ELBO= -384044.227949, dELBO=0.030391 </span>
<span class="co">#&gt; iter = 7, ELBO= -375892.948265, dELBO=0.021225 </span>
<span class="co">#&gt; iter = 8, ELBO= -370736.090977, dELBO=0.013719 </span>
<span class="co">#&gt; iter = 9, ELBO= -367314.756584, dELBO=0.009228 </span>
<span class="co">#&gt; iter = 10, ELBO= -365042.010765, dELBO=0.006187 </span>
<span class="co">#&gt; iter = 11, ELBO= -363428.051646, dELBO=0.004421 </span>
<span class="co">#&gt; iter = 12, ELBO= -362239.146930, dELBO=0.003271 </span>
<span class="co">#&gt; iter = 13, ELBO= -361317.545283, dELBO=0.002544 </span>
<span class="co">#&gt; iter = 14, ELBO= -360585.560598, dELBO=0.002026 </span>
<span class="co">#&gt; iter = 15, ELBO= -359989.346188, dELBO=0.001653 </span>
<span class="co">#&gt; iter = 16, ELBO= -359497.531867, dELBO=0.001366 </span>
<span class="co">#&gt; iter = 17, ELBO= -359086.408375, dELBO=0.001144 </span>
<span class="co">#&gt; iter = 18, ELBO= -358739.978436, dELBO=0.000965 </span>
<span class="co">#&gt; iter = 19, ELBO= -358445.622555, dELBO=0.000821 </span>
<span class="co">#&gt; iter = 20, ELBO= -358194.067552, dELBO=0.000702 </span>
<span class="co">#&gt; iter = 21, ELBO= -357977.844218, dELBO=0.000604 </span>
<span class="co">#&gt; iter = 22, ELBO= -357791.173993, dELBO=0.000521 </span>
<span class="co">#&gt; iter = 23, ELBO= -357629.328675, dELBO=0.000452 </span>
<span class="co">#&gt; iter = 24, ELBO= -357488.521436, dELBO=0.000394 </span>
<span class="co">#&gt; iter = 25, ELBO= -357365.610813, dELBO=0.000344 </span>
<span class="co">#&gt; iter = 26, ELBO= -357258.018654, dELBO=0.000301 </span>
<span class="co">#&gt; iter = 27, ELBO= -357163.581148, dELBO=0.000264 </span>
<span class="co">#&gt; iter = 28, ELBO= -357080.491309, dELBO=0.000233 </span>
<span class="co">#&gt; iter = 29, ELBO= -357007.218099, dELBO=0.000205 </span>
<span class="co">#&gt; iter = 30, ELBO= -356942.466750, dELBO=0.000181</span>
<span class="co">#&gt; 2023-07-28 09:35:56 : ***** Finish unwanted variation removal, 0.057 mins elapsed.</span>
<span class="co">#&gt; ******Sort the results into a Seurat object...</span>
<span class="co">#&gt; 2023-07-28 09:35:56 : ***** Finish results arrangement, 0.003 mins elapsed.</span>
<span class="va">seuInt</span>
<span class="co">#&gt; An object of class Seurat </span>
<span class="co">#&gt; 197 features across 1925 samples within 1 assay </span>
<span class="co">#&gt; Active assay: FAST (197 features, 0 variable features)</span>
<span class="co">#&gt;  3 dimensional reductions calculated: profast, harmony, position</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="visualization">Visualization<a class="anchor" aria-label="anchor" href="#visualization"></a>
</h2>
<p>First, user can choose a beautiful color schema using <code><a href="https://rdrr.io/pkg/PRECAST/man/chooseColors.html" class="external-link">chooseColors()</a></code> in the R package <code>PRECAST</code>.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cols_cluster</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/PRECAST/man/chooseColors.html" class="external-link">chooseColors</a></span><span class="op">(</span>palettes_name <span class="op">=</span> <span class="st">"Nature 10"</span>, n_colors <span class="op">=</span> <span class="fl">7</span>, plot_colors <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p><img src="FASTsimu_files/figure-html/unnamed-chunk-17-1.png" width="700"></p>
<p>Then, we plot the spatial scatter plot for clusters using the function <code><a href="https://rdrr.io/pkg/PRECAST/man/SpaPlot.html" class="external-link">SpaPlot()</a></code> in the R package <code>PRECAST</code>.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p12</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/PRECAST/man/SpaPlot.html" class="external-link">SpaPlot</a></span><span class="op">(</span><span class="va">seuInt</span>, item <span class="op">=</span> <span class="st">"cluster"</span>, batch <span class="op">=</span> <span class="cn">NULL</span>, point_size <span class="op">=</span> <span class="fl">1</span>, cols <span class="op">=</span> <span class="va">cols_cluster</span>, combine <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="va">p12</span></code></pre></div>
<p><img src="FASTsimu_files/figure-html/unnamed-chunk-18-1.png" width="768"> Users can re-plot the above figures for specific need by returning a ggplot list object. For example, we plot the spatial heatmap using a common legend by using the function <code><a href="https://rdrr.io/pkg/PRECAST/man/drawFigs.html" class="external-link">drawFigs()</a></code> in the R package <code>PRECAST</code>.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/PRECAST/man/SpaPlot.html" class="external-link">SpaPlot</a></span><span class="op">(</span><span class="va">seuInt</span>, item <span class="op">=</span> <span class="st">"cluster"</span>, title_name<span class="op">=</span> <span class="st">'Section'</span>,batch <span class="op">=</span> <span class="cn">NULL</span>, point_size <span class="op">=</span> <span class="fl">1</span>, cols <span class="op">=</span> <span class="va">cols_cluster</span>, combine <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/PRECAST/man/drawFigs.html" class="external-link">drawFigs</a></span><span class="op">(</span><span class="va">pList</span>, layout.dim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3</span><span class="op">)</span>, common.legend <span class="op">=</span> <span class="cn">TRUE</span>, legend.position <span class="op">=</span> <span class="st">"right"</span>, align <span class="op">=</span> <span class="st">"hv"</span><span class="op">)</span></code></pre></div>
<p><img src="FASTsimu_files/figure-html/unnamed-chunk-19-1.png" width="864"></p>
<p>We use the function <code><a href="https://rdrr.io/pkg/PRECAST/man/AddUMAP.html" class="external-link">AddUMAP()</a></code> in the R package <code>PRECAST</code> to obtain the three-dimensional components of UMAP using the aligned embeddings in the reduction <code>harmony</code>.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">seuInt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/PRECAST/man/AddUMAP.html" class="external-link">AddUMAP</a></span><span class="op">(</span><span class="va">seuInt</span>, n_comp<span class="op">=</span><span class="fl">3</span>, reduction <span class="op">=</span> <span class="st">'harmony'</span>, assay <span class="op">=</span> <span class="st">'FAST'</span><span class="op">)</span>
<span class="va">seuInt</span>
<span class="co">#&gt; An object of class Seurat </span>
<span class="co">#&gt; 197 features across 1925 samples within 1 assay </span>
<span class="co">#&gt; Active assay: FAST (197 features, 0 variable features)</span>
<span class="co">#&gt;  4 dimensional reductions calculated: profast, harmony, position, UMAP3</span></code></pre></div>
<p>We plot the spatial tNSE RGB plot to illustrate the performance in extracting features.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p13</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/PRECAST/man/SpaPlot.html" class="external-link">SpaPlot</a></span><span class="op">(</span><span class="va">seuInt</span>, batch <span class="op">=</span> <span class="cn">NULL</span>, item <span class="op">=</span> <span class="st">"RGB_UMAP"</span>, point_size <span class="op">=</span> <span class="fl">1</span>, combine <span class="op">=</span> <span class="cn">FALSE</span>, text_size <span class="op">=</span> <span class="fl">15</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/PRECAST/man/drawFigs.html" class="external-link">drawFigs</a></span><span class="op">(</span><span class="va">p13</span>, layout.dim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3</span><span class="op">)</span>, common.legend <span class="op">=</span> <span class="cn">TRUE</span>, legend.position <span class="op">=</span> <span class="st">"right"</span>, align <span class="op">=</span> <span class="st">"hv"</span><span class="op">)</span></code></pre></div>
<p><img src="FASTsimu_files/figure-html/unnamed-chunk-21-1.png" width="864"> We use the function <code>AddUTSNE()</code> in the R package <code>PRECAST</code> to obtain the two-dimensional components of UMAP using the aligned embeddings in the reduction <code>harmony</code>, and plot the tSNE plot based on the extracted features to check the performance of integration.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">seuInt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/PRECAST/man/AddTSNE.html" class="external-link">AddTSNE</a></span><span class="op">(</span><span class="va">seuInt</span>, n_comp <span class="op">=</span> <span class="fl">2</span>, reduction <span class="op">=</span> <span class="st">'harmony'</span>, assay <span class="op">=</span> <span class="st">'FAST'</span><span class="op">)</span>
<span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/PRECAST/man/dimPlot.html" class="external-link">dimPlot</a></span><span class="op">(</span><span class="va">seuInt</span>, item <span class="op">=</span> <span class="st">"cluster"</span>, point_size <span class="op">=</span> <span class="fl">0.5</span>, font_family <span class="op">=</span> <span class="st">"serif"</span>, cols <span class="op">=</span> <span class="va">cols_cluster</span>,
    border_col <span class="op">=</span> <span class="st">"gray10"</span>,  legend_pos <span class="op">=</span> <span class="st">"right"</span><span class="op">)</span>  <span class="co"># Times New Roman</span>
<span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/PRECAST/man/dimPlot.html" class="external-link">dimPlot</a></span><span class="op">(</span><span class="va">seuInt</span>, item <span class="op">=</span> <span class="st">"batch"</span>, point_size <span class="op">=</span> <span class="fl">0.5</span>, font_family <span class="op">=</span> <span class="st">"serif"</span>, legend_pos <span class="op">=</span> <span class="st">"right"</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/pkg/PRECAST/man/drawFigs.html" class="external-link">drawFigs</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">p1</span>, <span class="va">p2</span><span class="op">)</span>, layout.dim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span>, legend.position <span class="op">=</span> <span class="st">"right"</span><span class="op">)</span></code></pre></div>
<p><img src="FASTsimu_files/figure-html/unnamed-chunk-22-1.png" width="864"></p>
</div>
<div class="section level2">
<h2 id="combined-differential-epxression-analysis">Combined differential epxression analysis<a class="anchor" aria-label="anchor" href="#combined-differential-epxression-analysis"></a>
</h2>
<p>Finally, we condut the combined differential expression analysis using the integrated log-normalized expression matrix saved in the <code>seuInt</code> object. The function <code><a href="https://satijalab.org/seurat/reference/FindAllMarkers.html" class="external-link">FindAllMarkers()</a></code> in the <code>Seurat</code> R package is ued to achieve this analysis.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dat_deg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindAllMarkers.html" class="external-link">FindAllMarkers</a></span><span class="op">(</span><span class="va">seuInt</span><span class="op">)</span>
<span class="co">#&gt; Calculating cluster 1</span>
<span class="co">#&gt; Calculating cluster 2</span>
<span class="co">#&gt; Calculating cluster 3</span>
<span class="co">#&gt; Calculating cluster 4</span>
<span class="co">#&gt; Calculating cluster 5</span>
<span class="co">#&gt; Calculating cluster 6</span>
<span class="co">#&gt; Calculating cluster 7</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span>
<span class="co">#&gt; Warning: package 'dplyr' was built under R version 4.1.3</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'dplyr'</span>
<span class="co">#&gt; The following objects are masked from 'package:stats':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     filter, lag</span>
<span class="co">#&gt; The following objects are masked from 'package:base':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     intersect, setdiff, setequal, union</span>
<span class="va">n</span> <span class="op">&lt;-</span> <span class="fl">5</span>
<span class="va">dat_deg</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">cluster</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/top_n.html" class="external-link">top_n</a></span><span class="op">(</span>n <span class="op">=</span> <span class="va">n</span>, wt <span class="op">=</span> <span class="va">avg_log2FC</span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">top10</span>
<span class="va">seuInt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html" class="external-link">ScaleData</a></span><span class="op">(</span><span class="va">seuInt</span><span class="op">)</span>
<span class="co">#&gt; Centering and scaling data matrix</span></code></pre></div>
<p>Plot dot plot of normalized expressions for each spatial domain identified by using the FAST embeddings.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">col_here</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#F2E6AB"</span>, <span class="st">"#9C0141"</span><span class="op">)</span> 
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span>
<span class="co">#&gt; Warning: package 'ggplot2' was built under R version 4.1.3</span>
<span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DotPlot.html" class="external-link">DotPlot</a></span><span class="op">(</span><span class="va">seuInt</span>, features<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/unname.html" class="external-link">unname</a></span><span class="op">(</span><span class="va">top10</span><span class="op">$</span><span class="va">gene</span><span class="op">)</span>, cols<span class="op">=</span><span class="va">col_here</span>, <span class="co">#  idents = ident_here,</span>
              col.min <span class="op">=</span> <span class="op">-</span><span class="fl">1</span>, col.max <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span><span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>face <span class="op">=</span> <span class="st">"italic"</span><span class="op">)</span><span class="op">)</span><span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"Domain"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="cn">NULL</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size<span class="op">=</span><span class="fl">12</span>, angle <span class="op">=</span> <span class="fl">25</span>, hjust <span class="op">=</span> <span class="fl">1</span>, family<span class="op">=</span><span class="st">'serif'</span><span class="op">)</span>,
                                      axis.text.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size<span class="op">=</span><span class="fl">12</span>, face<span class="op">=</span> <span class="st">"italic"</span>, family<span class="op">=</span><span class="st">'serif'</span><span class="op">)</span><span class="op">)</span>
<span class="va">p1</span></code></pre></div>
<p><img src="FASTsimu_files/figure-html/unnamed-chunk-24-1.png" width="864"></p>
<details><p><summary><strong>Session Info</strong></summary></p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; R version 4.1.2 (2021-11-01)</span>
<span class="co">#&gt; Platform: x86_64-w64-mingw32/x64 (64-bit)</span>
<span class="co">#&gt; Running under: Windows 10 x64 (build 22621)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt; [1] LC_COLLATE=Chinese (Simplified)_China.936 </span>
<span class="co">#&gt; [2] LC_CTYPE=Chinese (Simplified)_China.936   </span>
<span class="co">#&gt; [3] LC_MONETARY=Chinese (Simplified)_China.936</span>
<span class="co">#&gt; [4] LC_NUMERIC=C                              </span>
<span class="co">#&gt; [5] LC_TIME=Chinese (Simplified)_China.936    </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] parallel  stats     graphics  grDevices utils     datasets  methods  </span>
<span class="co">#&gt; [8] base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt; [1] ggplot2_3.4.1      dplyr_1.0.9        sp_1.5-0           SeuratObject_4.1.0</span>
<span class="co">#&gt; [5] Seurat_4.1.1       PRECAST_1.6.1      FAST_1.2           gtools_3.9.2.2    </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;   [1] rappdirs_0.3.3              scattermore_0.8            </span>
<span class="co">#&gt;   [3] ggthemes_4.2.4              GiRaF_1.0.1                </span>
<span class="co">#&gt;   [5] ragg_1.2.2                  tidyr_1.2.0                </span>
<span class="co">#&gt;   [7] bit64_4.0.5                 knitr_1.37                 </span>
<span class="co">#&gt;   [9] irlba_2.3.5                 DelayedArray_0.20.0        </span>
<span class="co">#&gt;  [11] data.table_1.14.2           rpart_4.1.16               </span>
<span class="co">#&gt;  [13] KEGGREST_1.34.0             RCurl_1.98-1.6             </span>
<span class="co">#&gt;  [15] generics_0.1.2              BiocGenerics_0.40.0        </span>
<span class="co">#&gt;  [17] org.Mm.eg.db_3.14.0         ScaledMatrix_1.2.0         </span>
<span class="co">#&gt;  [19] cowplot_1.1.1               RSQLite_2.2.10             </span>
<span class="co">#&gt;  [21] RANN_2.6.1                  future_1.26.1              </span>
<span class="co">#&gt;  [23] bit_4.0.4                   spatstat.data_3.0-0        </span>
<span class="co">#&gt;  [25] xml2_1.3.3                  httpuv_1.6.5               </span>
<span class="co">#&gt;  [27] SummarizedExperiment_1.24.0 assertthat_0.2.1           </span>
<span class="co">#&gt;  [29] viridis_0.6.2               xfun_0.29                  </span>
<span class="co">#&gt;  [31] hms_1.1.1                   jquerylib_0.1.4            </span>
<span class="co">#&gt;  [33] evaluate_0.15               promises_1.2.0.1           </span>
<span class="co">#&gt;  [35] fansi_1.0.4                 progress_1.2.2             </span>
<span class="co">#&gt;  [37] dbplyr_2.1.1                igraph_1.3.5               </span>
<span class="co">#&gt;  [39] DBI_1.1.2                   htmlwidgets_1.5.4          </span>
<span class="co">#&gt;  [41] spatstat.geom_2.4-0         stats4_4.1.2               </span>
<span class="co">#&gt;  [43] purrr_0.3.4                 ellipsis_0.3.2             </span>
<span class="co">#&gt;  [45] RSpectra_0.16-1             ggpubr_0.4.0               </span>
<span class="co">#&gt;  [47] backports_1.4.1             DR.SC_3.2                  </span>
<span class="co">#&gt;  [49] insight_0.17.0              biomaRt_2.50.3             </span>
<span class="co">#&gt;  [51] deldir_1.0-6                sparseMatrixStats_1.6.0    </span>
<span class="co">#&gt;  [53] MatrixGenerics_1.6.0        vctrs_0.6.1                </span>
<span class="co">#&gt;  [55] SingleCellExperiment_1.16.0 Biobase_2.54.0             </span>
<span class="co">#&gt;  [57] ROCR_1.0-11                 abind_1.4-5                </span>
<span class="co">#&gt;  [59] cachem_1.0.6                withr_2.5.0                </span>
<span class="co">#&gt;  [61] progressr_0.10.1            sctransform_0.3.3          </span>
<span class="co">#&gt;  [63] prettyunits_1.1.1           mclust_5.4.10              </span>
<span class="co">#&gt;  [65] goftest_1.2-3               cluster_2.1.2              </span>
<span class="co">#&gt;  [67] lazyeval_0.2.2              crayon_1.5.1               </span>
<span class="co">#&gt;  [69] pkgconfig_2.0.3             labeling_0.4.2             </span>
<span class="co">#&gt;  [71] GenomeInfoDb_1.30.1         nlme_3.1-155               </span>
<span class="co">#&gt;  [73] vipor_0.4.5                 nnet_7.3-17                </span>
<span class="co">#&gt;  [75] rlang_1.1.0                 globals_0.15.0             </span>
<span class="co">#&gt;  [77] lifecycle_1.0.3             miniUI_0.1.1.1             </span>
<span class="co">#&gt;  [79] filelock_1.0.2              BiocFileCache_2.2.1        </span>
<span class="co">#&gt;  [81] rsvd_1.0.5                  rprojroot_2.0.3            </span>
<span class="co">#&gt;  [83] polyclip_1.10-0             matrixStats_0.62.0         </span>
<span class="co">#&gt;  [85] lmtest_0.9-40               Matrix_1.4-0               </span>
<span class="co">#&gt;  [87] carData_3.0-5               zoo_1.8-10                 </span>
<span class="co">#&gt;  [89] beeswarm_0.4.0              ggridges_0.5.3             </span>
<span class="co">#&gt;  [91] png_0.1-7                   viridisLite_0.4.1          </span>
<span class="co">#&gt;  [93] bitops_1.0-7                KernSmooth_2.23-20         </span>
<span class="co">#&gt;  [95] Biostrings_2.62.0           blob_1.2.2                 </span>
<span class="co">#&gt;  [97] DelayedMatrixStats_1.16.0   stringr_1.4.0              </span>
<span class="co">#&gt;  [99] parallelly_1.32.0           spatstat.random_2.2-0      </span>
<span class="co">#&gt; [101] rstatix_0.7.0               S4Vectors_0.32.3           </span>
<span class="co">#&gt; [103] ggsignif_0.6.3              beachmat_2.10.0            </span>
<span class="co">#&gt; [105] scales_1.2.1                memoise_2.0.1              </span>
<span class="co">#&gt; [107] magrittr_2.0.3              plyr_1.8.7                 </span>
<span class="co">#&gt; [109] ica_1.0-2                   zlibbioc_1.40.0            </span>
<span class="co">#&gt; [111] compiler_4.1.2              RColorBrewer_1.1-3         </span>
<span class="co">#&gt; [113] fitdistrplus_1.1-8          cli_3.2.0                  </span>
<span class="co">#&gt; [115] XVector_0.34.0              listenv_0.8.0              </span>
<span class="co">#&gt; [117] patchwork_1.1.1             pbapply_1.5-0              </span>
<span class="co">#&gt; [119] MASS_7.3-55                 mgcv_1.8-39                </span>
<span class="co">#&gt; [121] tidyselect_1.1.2            stringi_1.7.6              </span>
<span class="co">#&gt; [123] textshaping_0.3.6           highr_0.9                  </span>
<span class="co">#&gt; [125] yaml_2.3.6                  BiocSingular_1.10.0        </span>
<span class="co">#&gt; [127] ggrepel_0.9.1               grid_4.1.2                 </span>
<span class="co">#&gt; [129] sass_0.4.1                  tools_4.1.2                </span>
<span class="co">#&gt; [131] future.apply_1.9.0          rstudioapi_0.13            </span>
<span class="co">#&gt; [133] gridExtra_2.3               farver_2.1.1               </span>
<span class="co">#&gt; [135] Rtsne_0.16                  digest_0.6.29              </span>
<span class="co">#&gt; [137] rgeos_0.5-9                 FNN_1.1.3.1                </span>
<span class="co">#&gt; [139] shiny_1.7.1                 Rcpp_1.0.10                </span>
<span class="co">#&gt; [141] GenomicRanges_1.46.1        car_3.0-12                 </span>
<span class="co">#&gt; [143] broom_0.7.12                scuttle_1.4.0              </span>
<span class="co">#&gt; [145] performance_0.8.0           later_1.3.0                </span>
<span class="co">#&gt; [147] harmony_0.1.0               RcppAnnoy_0.0.19           </span>
<span class="co">#&gt; [149] org.Hs.eg.db_3.14.0         httr_1.4.3                 </span>
<span class="co">#&gt; [151] AnnotationDbi_1.56.2        colorspace_2.1-0           </span>
<span class="co">#&gt; [153] XML_3.99-0.9                fs_1.5.2                   </span>
<span class="co">#&gt; [155] tensor_1.5                  reticulate_1.25            </span>
<span class="co">#&gt; [157] IRanges_2.28.0              splines_4.1.2              </span>
<span class="co">#&gt; [159] uwot_0.1.11                 spatstat.utils_3.0-1       </span>
<span class="co">#&gt; [161] pkgdown_2.0.6               scater_1.25.1              </span>
<span class="co">#&gt; [163] plotly_4.10.0               systemfonts_1.0.4          </span>
<span class="co">#&gt; [165] xtable_1.8-4                jsonlite_1.8.0             </span>
<span class="co">#&gt; [167] R6_2.5.1                    pillar_1.9.0               </span>
<span class="co">#&gt; [169] htmltools_0.5.2             mime_0.12                  </span>
<span class="co">#&gt; [171] glue_1.6.2                  fastmap_1.1.0              </span>
<span class="co">#&gt; [173] BiocParallel_1.28.3         BiocNeighbors_1.12.0       </span>
<span class="co">#&gt; [175] codetools_0.2-18            utf8_1.2.3                 </span>
<span class="co">#&gt; [177] lattice_0.20-45             bslib_0.3.1                </span>
<span class="co">#&gt; [179] spatstat.sparse_2.1-1       tibble_3.2.1               </span>
<span class="co">#&gt; [181] curl_4.3.2                  ggbeeswarm_0.6.0           </span>
<span class="co">#&gt; [183] leiden_0.4.2                limma_3.50.1               </span>
<span class="co">#&gt; [185] survival_3.2-13             CompQuadForm_1.4.3         </span>
<span class="co">#&gt; [187] rmarkdown_2.11              desc_1.4.0                 </span>
<span class="co">#&gt; [189] munsell_0.5.0               GenomeInfoDbData_1.2.7     </span>
<span class="co">#&gt; [191] reshape2_1.4.4              gtable_0.3.3               </span>
<span class="co">#&gt; [193] spatstat.core_2.4-4</span></code></pre></div>
</details>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Wei Liu, Xiao Zhang, Jin Liu.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
